### ISSUE TITLE
ModSpec test encoding with sub-tests and steps

### ISSUE THREAD

**Issue by ronaldtse** [metanorma/metanorma-ogc#274]
_This ticket documents the implementation of ModSpec test encoding changes for metanorma/ogc-citygml-xmi#48._

# Original issue

## Case 1

Original:
<img width="841" alt="Screenshot 2021-09-01 at 2 31 43 PM" src="https://user-images.githubusercontent.com/11865/131623155-86ffce7f-8ba1-4da2-91f9-fa495c887b63.png">

Now:

<img width="825" alt="Screenshot 2021-09-01 at 2 31 35 PM" src="https://user-images.githubusercontent.com/11865/131623054-7b813161-cf83-4129-8c69-351cae9748c5.png">

## Case 2

Original:
<img width="839" alt="Screenshot 2021-09-01 at 2 44 15 PM" src="https://user-images.githubusercontent.com/11865/131624502-c6fcbbe6-facb-4ecc-ae40-ea10a8a7ba16.png">

New:
<img width="813" alt="Screenshot 2021-09-01 at 2 44 27 PM" src="https://user-images.githubusercontent.com/11865/131624526-19e8452a-07b4-4cb7-981f-631d1e642056.png">

## What exactly is this line?

* It's not a test "condition" as in relation to ModSpec (OGC 08-131r3).
* It looks like this line is part of the "Test Method"?

## What does the ModSpec say?

The relevant portions of are listed below:

<img width="558" alt="Screenshot 2021-09-01 at 2 40 30 PM" src="https://user-images.githubusercontent.com/11865/131624073-43804af7-5beb-48bb-ad0c-bc3c908f4210.png">

<img width="567" alt="Screenshot 2021-09-01 at 2 39 51 PM" src="https://user-images.githubusercontent.com/11865/131623990-07be08fa-ae28-45c0-ad68-88a7e4985b4c.png">

<img width="577" alt="Screenshot 2021-09-01 at 2 40 51 PM" src="https://user-images.githubusercontent.com/11865/131624111-72a3c506-df21-48d0-b57c-57394e60bcb0.png">

<img width="584" alt="Screenshot 2021-09-01 at 2 40 06 PM" src="https://user-images.githubusercontent.com/11865/131624020-e7a26e36-6a95-4685-a2a9-6acf729832c6.png">

## Next steps

I think we need to decide whether a sub-test is an actually self-sufficient test, i.e. whether it can be tested independently.

For example, the first example writes:
> If any ADE classes or properties are included in the Generics Package:
> A: Validate that the Implementation Specification conforms with the ADE Requirements Class

What this means is that "If" the stated condition happens (at the implementation level, not at the specification level), then the sub-test A applies.

However, if we have B, C, it also means that the "If" condition applies to all of those. There are many Abstract Tests with multiple sub-tests that require the same condition.

According to ModSpec, which disallows conditions on specific Tests (conditions are only applied to Classes), I would think that the "If" condition is a necessary condition of the sub-test.

i.e. Such "If" and "For" statements are actually part of the sub-test. 

In this case, the sub-test "A" would be:
> A: If any ADE classes or properties are included in the Generics Package, validate that the Implementation Specification conforms with the ADE Requirements Class

And if there is a sub-test "B", it would be:
> B: If any ADE classes or properties are included in the Generics Package, {do this}

## TLDR

TLDR what I'm trying to say is:
* a sub-test inside an Abstract Test (or any Test) *should not* have a shared condition that is stated outside allowed structure. The shared condition should be either   
  * internalised into individual sub-tests;
  * made into a separate Test Class; or
  * ModSpec needs to be changed to allow grouping sub-tests with a condition within a Test

# Original enquiry to OGC

In the CityGML 3.0 standard, some abstract tests look like this:
[Note that an “abstract test” is just a short form for a “concrete conformance test” (which the term “abstract test” is not defined in ModSpec).]

e.g.

Sample requirement:
<img width="895" alt="Screenshot 2021-09-03 at 11 46 15 AM" src="https://user-images.githubusercontent.com/11865/132607180-5114d13f-00b3-4ab7-a064-539063baa6f9.png">

Corresponding test for the sample requirement:
<img width="899" alt="Screenshot 2021-09-03 at 12 01 46 PM" src="https://user-images.githubusercontent.com/11865/132607136-97f69456-fcde-4e67-aaf7-46916af58d12.png">

————
Problem 1: Do conformance tests allow parts / steps?

Let me first reproduce the structural definition of a “Conformance test” according to ModSpec:

<img width="563" alt="Screenshot 2021-09-03 at 12 29 13 PM" src="https://user-images.githubusercontent.com/11865/132607120-698da471-5c7d-416e-b6f6-98126dbebdb0.png">

The ModSpec document does actually contain errors that require fixing, so we have to take its intent into consideration. In CityGML 3.0 abstract tests, they often use “A”, “B”, “C”, etc. Presumably, these are either “parts” within a test, or “sequential steps” of the test.

A reading of ModSpec indicates that both “parts” and “sequential steps” are allowable?

- supporting parts: ModSpec 4.3 says "conformance test case: test for a particular requirement or a set of related requirements. NOTE When no ambiguity, the word ― “case” may be omitted. i.e. conformance test is the same as conformance.” This NOTE implies that a “conformance test” may or may not be a test case, i.e. a “conformance test” may contain multiple “conformance test cases”.

- supporting sequential steps: ModSpec C.11 says “In this standard, conditional tests, based on a requirement with some precondition, have as part of their execution the task of creating the precondition in which they will be tested”. This implies that a test case can contain multiple steps.

- supporting none: ModSpec Annex C provides a UML diagram which does not provide for parts or steps, and the C.11 UML definition supports neither too.

What is the intended model here?

Here’s a stab at it that allows both parts and steps:

```
// A Conformance Test inherits from Conformance Test Case
class ConformanceTest > ConformanceTestCase {
  testPurpose: String
  testType: TestType
  reference: String[0..*]
  requirements: Requirement[1..*]
  parts: ConformanceTestCase[0..*]
}

// A Conformance Test can contain multiple parts (Test Cases)
class ConformanceTestCase {
  testMethod: ConformanceTestStep[1..*]
}

// A Conformance Test Case can contain multiple Steps
class ConformanceTestStep {
  steps: String[1..*]
}

```

Problem 2: Conformance tests and conditions / parameters.

According to ModSpec 6.5.1:

"**A Conformance class may be parameterized**. This means that the class’s tests depend on some parameter that must be defined before the tests can be executed. For example, if a XYZ conformance class needs to specify a data format such as GML or KML to be tested, then XYZ(GML) is XYZ using GML, and XYZ(KML) is XYZ using KML. **Because the parameters choose which requirements will be tested, two conformance classes with distinct parameters should be considered as distinct conformance classes.**”

(red bold text in original styling)

Looking at the screenshots above for CityGML 3.0, the line “If any ADE classes or properties are included in the Generics Package:” line is “like” a condition or parameter. We have seen “If…” and “For all…” patterns in abstract tests.

There are a few problems here:

1. 6.5.1 says Only conformance classes can be parameterised, not conformance tests.
2. 6.5.1 also says that parameterisation requires the conformance classes to be distinct, i.e. XYZ(GML) and XYZ(KML) are separate conformance classes.
3. Recall that ModSpec, C.11 (Conformance tests) also specifies the following:
“In this standard, conditional tests, based on a requirement with some precondition, have as part of their execution the task of creating the precondition in which they will be tested”. This implies that a conformance test can include preconditions, as long as the precondition is created as a step within the conformance test.

So this is more of a matter of “how” to fit these abstract tests into ModSpec models.

There are a few options:

0. (This may be needed in any case) Apply the same parameterisation principles to Conformance Tests in addition to Conformance Classes.

1. Treat the quoted line (“If …”) not as a condition or parameter, but internalised by individual parts, i.e.

- Abstract Test 18A would be “If any ADE classes …, validate that the Implementation Specification…”
- Abstract Test 18B would be “If any ADE classes …, {original B content}…”, etc.

2. Treat the quoted line (“If …”) as a “precondition” aka the first step in a sequence of steps as specified in C.11, i.e.

- Step 1 “Check if there are any ADE classes …, if none, return."
- Step 2 “Validate that all classes from Step 1 conform with the Implementation Specification…” (content from 18A)
- Step 3 “Validate that all classes from Step 1 conform with …” etc. (content from 18B)

3. Allow hierarchical steps in a Conformance Test Case, and do 2 but in a hierarchical manner (i.e. “If …, do {step A}, {step B} and {step C}”).

———

Sorry for this long message but I think for the long term it is crucial to have a correct way of encoding ModSpec requirements. Thanks in advance!

**Comment by ronaldtse** [comment]
As suggested by @cmheazel , the OGC API document provides individual steps under "Test Method":

http://docs.opengeospatial.org/is/17-069r3/17-069r3.html#_general_tests

e.g.
<img width="770" alt="Screenshot 2021-09-07 at 10 57 51 PM" src="https://user-images.githubusercontent.com/11865/132366731-92f55a35-19a0-4683-9066-6bbc3d98d583.png">

<img width="768" alt="Screenshot 2021-09-07 at 10 57 33 PM" src="https://user-images.githubusercontent.com/11865/132366692-aeebf412-26f3-499f-bcc5-43e96140dd39.png">

Notice that in CityGML, the "Test Method" field is actually for indicating "Test Method Type" (e.g. "Manual Inspection") instead of describing the steps of testing.

**Comment by ronaldtse** [comment]
From @opoudjis :

We are going to realise this as follows, after discussion:

* The current "test method" can be marked up as a "test method type", since we have deliberately allowed component classes to be open ended:

````
[.component,class=Test method type]
=====
Manual Inspection
=====
```` 

* I have allowed components of requirements to be marked up with example delimiters as well as open block delimiters, so as to allow recursion (see above)
* We are introducing the component "step" as a nested component within the component "test method", in order to indicate the logical steps that are to be taken within a requirement test
  * The "step" components can be recursive
  * The steps of a test method can include conditional statements. Nesting of subsequent steps can be used to indicate the scope of conditional statements
  * Steps as components are only meaningful when contained within a test method; they are not meaningful as first-level requirements of a test
  * Steps of test methods are DISTINCT from Parts of requirements. A part of a requirement is a subrequirement, an independent requirement. A step is a part of a sequence of a test method of a test: it is not a subtest.
  * Markup in OGC to date has been inconsistent in this regard, but going forward, we are going to be differentiating steps in rendering. There are instances of Parts (A, B, C) of tests which will need to be marked up in Metanorma as steps instead. CityGML is such an instance.
  * Steps shall be rendered as numbered, nested lists

````
[.component,class=Test method type]
--
Manual Inspection
--

[.component,class=Test method]
=====

[.component,class=step]
======
For each UML class defined or referenced in the Tunnel Package:

[.component,class=step]
--
Validate that the Implementation Specification contains a data element which represents the same concept as that defined for the UML class.
--

[.component,class=step]
--
Validate that the data element has the same relationships with other elements as those defined for the UML class. Validate that those relationships have the same source, target, direction, roles, and multiplicies as those documented in the Conceptual Model.
--

[.component,class=step]
--
Validate that the data element has the same properties (attributes) as those specified for the UML class. Validate that those properties have the same name, definition, type, and multiplicity of those documented in the Conceptual Model.
--

[.component,class=step]
--
Validate that the properties of the data element include those of all superclasses of the UML class as documented in the Conceptual Model. Validate that those properties have the same name, definition, type, and multiplicity of those documented in the Conceptual Model
--

[.component,class=step]
--
Validate that the associations represented for the data element include those of all superclasses of the UML class as documented in the Conceptual Model. Validate that those representations have the same source, target, roles, and multiplicity of those documented in the Conceptual Model
--

[.component,class=step]
--
Validate that the Implementation Specification enforces all constraints imposed on the UML class by the Conceptual Model
--
======
=====
````

This renders as:

````
| Test method type | Manual inspection
| Test method      | 1. For each UML class defined or referenced in the Tunnel Package:
|                  |     a. Validate that the Implementation Specification contains a data element which represents the same concept as that defined for the UML class.
|                  |     b. Validate that the data element has the same relationships with other elements as those defined for the UML class. Validate that those relationships have the same source, target, direction, roles, and multiplicies as those documented in the Conceptual Model.
....
````

**Comment by ronaldtse** [comment]
1. CityGML uses the “Test Method” field to mean “the type of testing method”, filing it with the value “Manual Inspection”. In API-Features, it is either a single step or a series of steps.

2. In CityGML abstract tests, if the corresponding requirement contains parts (e.g. “A, B”), then the abstract test contains all parts (“A, B”). In API-Features, requirements that contain parts have abstract tests that do not necessarily correspond to those parts (in fact, the “Test Method” is a step tree with no regard for the parts). e.g. a requirement with A, B, C, can have a test that only tests for A (and does not mention A either).

Could we can move ahead with the following understanding?

- A test can have multiple Test Methods (Test Method A, Test Method B…). Each of these Test Methods may link to a requirement or a requirement part (A, B…).
- Each "Test Method” is a tree of steps.

**Comment by ronaldtse** [comment]
This understanding is now confirmed by OGC via @ogcscotts.

**Comment by ronaldtse** [comment]
This has been implemented in 16c996cc34a22b19d968b976b77d3e489ce0de15

**Comment by ronaldtse** [comment]
The resulting conceptual model of a ModSpec Test (conformance test or an abstract test).

```
// A Conformance Test includes potentially multiple Test Methods
// A Conformance Test Method is considered a "part" of the test (A, B...)
class ConformanceTest {
  testPurpose: String
  testType: TestType
  reference: RichText[0..*]
  linkToSpecification: RequirementOrRequirementPart[1..*]
  testMethod: ConformanceTestMethod[0..*]
}

// A Test Method can contain multiple steps
// A Test Method can be linked to a particular Requirement or Requirement Part
class ConformanceTestMethod {
  linkToSpecification: RequirementOrRequirementPart[0..*]
  steps: ConformanceTestStep[1..*]
}

// A Test Step can contain inner Test Steps
class ConformanceTestStep {
  content: RichText[1..*]
  inner: ConformanceTestStep[0..*]
}
```

**Comment by ronaldtse** [comment]
We need to update the modspec LutaML copy in metanorma-requirements-models.

**Comment by ronaldtse** [comment]
Need documentation too.

**Comment by ronaldtse** [comment]
Done:
* published on metanorma.org (https://github.com/metanorma/metanorma.org/pull/479)
* ModSpec model updated at https://github.com/metanorma/metanorma-requirements-models/commit/84169dcba4f700bd9eee2babfa26eb2a8629675f
